RileyLinkKitTests/PumpOpsSynchronousTests.swift

```diff
@@ -0,0 +1,480 @@
+        let outOfOrderPumpEvents: [PumpEvent] = [batteryEvent2017, batteryEvent2007]
+        
+        let (events, _) = sut.convertPumpEventToTimestampedEvents(pumpEvents: outOfOrderPumpEvents, startDate: Date.distantPast, checkDate: datePast2017, pumpModel: pumpModel)
+        
+        XCTAssertEqual(events.count, 1)
+    }
+    
+    func loadTestWithPumpModel(_ newPumpModel: PumpModel) {
+        pumpModel = newPumpModel
+        loadSUT()
+    }
+    
+    func testMultipleBolusEvents() {
+        loadTestWithPumpModel(.Model522)
+        
+        let events = [createBolusEvent2010(), createBolusEvent2009()]
+        
+        let (timeStampedEvents, _) = sut.convertPumpEventToTimestampedEvents(pumpEvents: events, startDate: Date.distantPast, checkDate: datePast2017, pumpModel: pumpModel)
+        
+        XCTAssertEqual(timeStampedEvents.count, 2)
+    }
+    
+    func testMultipleBolusEventsContainsFirstBolus() {
+        loadTestWithPumpModel(.Model522)
+        
+        let events = [createBolusEvent2010(), createBolusEvent2009()]
+        
+        let (timeStampedEvents, _) = sut.convertPumpEventToTimestampedEvents(pumpEvents: events, startDate: Date.distantPast, checkDate: datePast2017, pumpModel: pumpModel)
+        
+        assertArray(timeStampedEvents, containsPumpEvent: createBolusEvent2010())
+    }
+    
+    func testMultipleBolusEventsContainsSecondBolus() {
+        loadTestWithPumpModel(.Model522)
+        
+        let events = [createBolusEvent2010(), createBolusEvent2009()]
+        
+        let (timeStampedEvents, _) = sut.convertPumpEventToTimestampedEvents(pumpEvents: events, startDate: Date.distantPast, checkDate: datePast2017, pumpModel: pumpModel)
+        
+        assertArray(timeStampedEvents, containsPumpEvent: createBolusEvent2009())
+    }
+    
+    func testNonMutableBolusEventFor522() {
+        // device that can have out of order events
+        loadTestWithPumpModel(.Model522)
+        // 2009-07-31 09:00:00 +0000
+        // 120 minute duration
+        let regularBolus = BolusNormalPumpEvent(availableData: Data(hexadecimalString: "010080048000240009a24a1510")!, pumpModel: pumpModel)!
+        
+        let hourAfterRegularBolusDate = regularBolus.timestamp.date!.addingTimeInterval(TimeInterval(minutes: 60))
+        let events:[PumpEvent] = [regularBolus]
+        
+        let (timeStampedEvents, _) = sut.convertPumpEventToTimestampedEvents(pumpEvents: events, startDate: Date.distantPast, checkDate: hourAfterRegularBolusDate, pumpModel: pumpModel)
+        
+        // It should be counted for IoB (how to measure)
+        XCTAssertTrue(array(timeStampedEvents, containsPumpEvent: regularBolus))
+    }
+    
+    func testOutOfOrderEventFor522() {
+        // device that can have out of order events
+        loadTestWithPumpModel(.Model522)
+        
+        let data = Data(hexadecimalString:"338c4055145d1000")!
+        
+        // Create event like Temp Bolus (priming pump
```

Comment:
Square wave bolus on 522 is one of the out of order events.  Temp bolus isn't a thing. ;)